core/trino-main/src/main/java/io/trino/operator/aggregation/DistinctAccumulatorFactory.java

```diff
@@ -193,7 +202,12 @@ private DistinctGroupedAccumulator(
                     Optional.empty(),
                     joinCompiler,
                     blockTypeOperators,
-                    UpdateMemory.NOOP);
+                    () -> {
+                        // enforce task memory limits for fast throw
+                        updateMemory.update();
```

Comment:
Actually, I think this is wrong. `updateMemory.update()` will use `setBytes` and caller is obligated to wait for future to be done. This is the contract of memory management. If we break it, then OOM can still happen when there are many queries which put memory pressure.

I think `void addInput(GroupByIdBlock groupIdsBlock, Page page, Optional<Block> mask);` should just return `Work` or we could have an approach like in https://github.com/trinodb/trino/issues/460

Reply:
any updates?

Reply:
Discussion on-going in Slack